hypothesis-python/tests/cover/test_lookup.py

```diff
@@ -655,3 +656,26 @@ class TreeForwardRefs(typing.NamedTuple):
 @given(st.builds(TreeForwardRefs))
 def test_resolves_forward_references_outside_annotations(t):
     assert isinstance(t, TreeForwardRefs)
+
+
+def constructor(a: str = None):
```

Comment:
Well, this is a really hard question! 

Because, it works differently in different type checkers and even with different options in a single type checker. `mypy` with its `no-implicit-optional` is a good example.

From the typing perspective, it is as wrong as `a: str = 1`, but we know the context behind it, so it does not seem as scary for us. 

Maybe we can think about using `Union[annotation, default]`? It is more generic and can be less (or more 😄 ) confusing. 

Another idea is to add a new parameter for `builds` and `from_type` something like `include_defaults` which can do what I have describe above.

Reply:
> Because, it works differently in different type checkers and even with different options in a single type checker. `mypy` with its `no-implicit-optional` is a good example.

My specific motivation is that we follow the behaviour of `typing.get_type_hints()` - the "most standard" type checker - *unless* it's on a class and we override from an explictly defined `__signature__`.

Unfortunately we can't add a parameter to `st.builds()` because all the `*args, **kwargs` may be in use; and we can't add it to `st.from_type()` because that would only apply when it falls back to `st.builds()` (in which case...).

Treating `infer` as meaning `just(default) | from_type(annotation)` is tempting, but fails on the semi-valid sentinel values that you identify below (and they're often transformed by some framework, so we can't in general rely on being allowed to explicitly pass the default value).

So `arg: T=None` is a specific special case, and I argue we should treat it as such.

Reply:
Great point!